api/src/org/labkey/api/module/ModuleLoader.java

```diff
@@ -629,6 +637,34 @@ private void checkForRenamedModules()
             });
     }
 
+    /** Create a file that indicates the server is in the midst of the upgrade process. Refuse to start up
+     * if a previous startup left the lock file in place. */
+    private File createLockFile(File modulesDir) throws ConfigurationException
+    {
+        File result = new File(modulesDir.getParentFile(), "labkeyUpgradeLockFile");
+        if (result.exists())
+        {
+            if (AppProps.getInstance().isDevMode())
+            {
+                _log.warn("Lock file " + FileUtil.getAbsoluteCaseSensitiveFile(result) + " already exists - a previous upgrade attempt may have left the server in an indeterminate state.");
+                _log.warn("Bravely continuing because this server is running in Dev mode.");
+            }
+            else
+            {
+                throw new ConfigurationException("Lock file " + FileUtil.getAbsoluteCaseSensitiveFile(result) + " already exists - a previous upgrade attempt may have left the server in an indeterminate state. Proceed with extreme caution as the database may not be properly upgraded. To continue, delete file and restart Tomcat.");
+            }
+        }
+        try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(result), StringUtilsLabKey.DEFAULT_CHARSET)))
+        {
+            writer.write("LabKey instance beginning initialization at " + new Date());

```

Comment:
We try not to output `Date.toString()`...
```suggestion
            writer.write("LabKey instance beginning initialization at " + DateUtil.formatDateTimeISO8601(new Date()));

```